File: apps/api/internal/pbl/paper_layout.go
package pbl

import (
	"fmt"
	"math"
	"strings"
	"unicode/utf8"
)

// PaperLayout contains inert vector primitives, never model-authored HTML/SVG.
// Coordinates are print millimetres, not measurements of the depicted object.
type PaperLayout struct {
	Format   string         `json:"format"`
	Title    string         `json:"title"`
	Notice   string         `json:"notice"`
	Elements []PaperElement `json:"elements"`
}
type PaperElement struct {
	Kind     string  `json:"kind"`
	X        float64 `json:"x"`
	Y        float64 `json:"y"`
	Width    float64 `json:"width"`
	Height   float64 `json:"height"`
	Text     string  `json:"text,omitempty"`
	FontSize float64 `json:"fontSize,omitempty"`
	Fill     string  `json:"fill,omitempty"`
	Dashed   bool    `json:"dashed,omitempty"`
}

func (p PaperLayout) Validate() error {
	if p.Format != "a4-portrait" || strings.TrimSpace(p.Title) == "" || utf8.RuneCountInString(p.Title) > 28 || strings.TrimSpace(p.Notice) == "" || utf8.RuneCountInString(p.Notice) > 60 {
		return fmt.Errorf("纸面成果需要A4竖向格式、28字内标题和60字内说明")
	}
	if strings.ContainsAny(p.Title+p.Notice, "\r\n") {
		return fmt.Errorf("纸面标题和说明须为单行")
	}
	if len(p.Elements) == 0 || len(p.Elements) > 120 {
		return fmt.Errorf("纸面成果须有1至120个图形或文字元素")
	}
	for i, e := range p.Elements {
		for _, v := range []float64{e.X, e.Y, e.Width, e.Height, e.FontSize} {
			if math.IsNaN(v) || math.IsInf(v, 0) {
				return fmt.Errorf("纸面元素%d坐标无效", i+1)
			}
		}
		height := e.Height
		if e.Kind == "text" {
			height = math.Max(height, e.FontSize*1.3)
		}
		if e.X < 12 || e.Y < 35 || e.Width < 0 || e.Height < 0 || e.X+e.Width > 198 || e.Y+height > 277 {
			return fmt.Errorf("纸面元素%d超出可打印区域：x12至198，y35至277", i+1)
		}
		if e.Fill != "" && e.Fill != "none" && e.Fill != "white" && e.Fill != "light" && e.Fill != "dark" {
			return fmt.Errorf("纸面元素%d填充色无效", i+1)
		}
		switch e.Kind {
		case "rect", "ellipse":
			if e.Width <= 0 || e.Height <= 0 {
				return fmt.Errorf("纸面图形%d需要正的宽高", i+1)
			}
		case "line":
			if e.Width == 0 && e.Height == 0 {
				return fmt.Errorf("纸面线段%d长度为零", i+1)
			}
		case "text":
			if e.FontSize < 3 || e.FontSize > 8 {
				return fmt.Errorf("纸面文字%d字号为%gmm，须在3至8mm之间", i+1, e.FontSize)
			}
			if strings.TrimSpace(e.Text) == "" || strings.ContainsAny(e.Text, "\r\n") {
				return fmt.Errorf("纸面文字%d须为非空单行，换行请拆成多个text元素", i+1)
			}
			width := 0.0
			for _, r := range e.Text {
				if r < 128 {
					width += e.FontSize * 0.65
				} else {
					width += e.FontSize
				}
			}
			if width > e.Width {
				return fmt.Errorf("纸面文字%d过长，请缩短或拆成多行元素", i+1)
			}
		default:
			return fmt.Errorf("纸面元素%d类型不支持，只允许rect/ellipse/line/text", i+1)
		}
	}
	return nil
}

func (p PaperLayout) Markdown() string {
	lines := []string{"# " + p.Title, p.Notice, "图形版请查看纸面预览。以下为页面文字记录："}
	for _, e := range p.Elements {
		if e.Kind == "text" {
			lines = append(lines, e.Text)
		}
	}
	return strings.Join(lines, "\n\n")
}

File: apps/lite-web/src/api/paperLayout.ts
export interface PaperElement {
  kind: "rect" | "ellipse" | "line" | "text";
  x: number; y: number; width: number; height: number;
  text?: string; fontSize?: number; fill?: "none" | "white" | "light" | "dark"; dashed?: boolean;
}
export interface PaperLayout { format: "a4-portrait"; title: string; notice: string; elements: PaperElement[] }

export function readPaperLayout(value: unknown): PaperLayout | null {
  if (!value || typeof value !== "object") return null;
  const p = value as PaperLayout;
  if (p.format !== "a4-portrait" || typeof p.title !== "string" || typeof p.notice !== "string" || !Array.isArray(p.elements) || !p.elements.length || p.elements.length > 120) return null;
  if (!p.elements.every(e => e && (e.fill === undefined || ["none","white","light","dark"].includes(e.fill)) && (e.dashed === undefined || typeof e.dashed === "boolean") && ["rect", "ellipse", "line", "text"].includes(e.kind) && [e.x,e.y,e.width,e.height].every(Number.isFinite) && e.x >= 12 && e.y >= 35 && e.width >= 0 && e.height >= 0 && e.x+e.width <= 198 && e.y+e.height <= 277 && (e.kind !== "text" || (typeof e.text === "string" && Number.isFinite(e.fontSize) && e.fontSize! >= 3 && e.fontSize! <= 8)))) return null;
  return p;
}

const escape = (s: string) => s.replace(/[&<>"']/g, c => ({"&":"&amp;","<":"&lt;",">":"&gt;",'"':"&quot;","'":"&#39;"})[c]!);
export function paperSVG(layout: PaperLayout): string {
  if (!readPaperLayout(layout)) throw new Error("纸面布局无效");
  // Text is always foreground; model array order does not control legibility.
  const ordered = [...layout.elements.filter(e => e.kind !== "text"), ...layout.elements.filter(e => e.kind === "text")];
  const shapes = ordered.map(e => {
    const fill = {none:"none",white:"#fff",light:"#f0f0ec",dark:"#273735"}[e.fill ?? "none"] ?? "none";
    const style = `fill="${fill}" stroke="#273735" stroke-width="0.4"${e.dashed ? ' stroke-dasharray="2 1.5"' : ""}`;
    switch (e.kind) {
      case "rect": return `<rect x="${e.x}" y="${e.y}" width="${e.width}" height="${e.height}" ${style}/>`;
      case "ellipse": return `<ellipse cx="${e.x+e.width/2}" cy="${e.y+e.height/2}" rx="${e.width/2}" ry="${e.height/2}" ${style}/>`;
      case "line": return `<line x1="${e.x}" y1="${e.y}" x2="${e.x+e.width}" y2="${e.y+e.height}" stroke="#273735" stroke-width="0.4"${e.dashed ? ' stroke-dasharray="2 1.5"' : ""}/>`;
      case "text": return `<text x="${e.x}" y="${e.y+e.fontSize!}" font-size="${e.fontSize}" fill="#273735" stroke="white" stroke-width="0.6" paint-order="stroke fill">${escape(e.text!)}</text>`;
    }
  }).join("");
  return `<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 210 297" role="img" aria-label="${escape(layout.title)}"><rect width="210" height="297" fill="white"/><g font-family="Noto Sans CJK SC,Microsoft YaHei,sans-serif"><text x="12" y="19" font-size="6" font-weight="700" fill="#273735">${escape(layout.title)}</text><text x="12" y="28" font-size="3" fill="#555">${escape(layout.notice)}</text><g id="paper-elements">${shapes}</g><line x1="12" y1="283" x2="198" y2="283" stroke="#aaa" stroke-width="0.2"/><text x="12" y="289" font-size="2.8" fill="#555">A4 · 图形为纸面示意，非实物尺寸</text></g></svg>`;
}

export function paperHTML(layout: PaperLayout): string {
  return `<!doctype html><html lang="zh-CN"><meta charset="utf-8"><title>${escape(layout.title)}</title><style>*{box-sizing:border-box}html,body{margin:0;background:#fff}.sheet{width:210mm;height:297mm}.sheet svg{display:block;width:100%;height:100%;overflow:visible}@page{size:A4 portrait;margin:0}@media print{.sheet{break-inside:avoid;break-after:avoid}}</style><body><main class="sheet">${paperSVG(layout)}</main></body></html>`;
}

export function downloadPaper(layout: PaperLayout) {
  const url = URL.createObjectURL(new Blob([paperHTML(layout)], {type:"text/html;charset=utf-8"}));
  const a = document.createElement("a"); a.href=url; a.download="纸面原型-A4.html"; a.click();
  setTimeout(() => URL.revokeObjectURL(url), 1000);
}
